src/stactools/viirs/metadata.py

```diff
@@ -0,0 +1,355 @@
+import ast
+import datetime
+import logging
+import os.path
+import warnings
+from dataclasses import dataclass
+from typing import Any, Callable, Dict, List, Optional, Tuple
+
+import fsspec
+import h5py
+import numpy as np
+import rasterio
+import shapely.geometry
+from dateutil import parser
+from lxml import etree
+from rasterio.errors import NotGeoreferencedWarning
+from shapely.geometry import Polygon
+from stactools.core.io import ReadHrefModifier
+from stactools.core.io.xml import XmlElement
+from stactools.core.projection import reproject_geom
+from stactools.core.utils import href_exists
+
+from stactools.viirs import constants, utils
+
+logger = logging.getLogger(__name__)
+
+
+class MissingElement(Exception):
+    """An expected element is missing from the XML file"""
+
+
+@dataclass
+class Metadata:
+    """Structure to hold values from a metadata XML file or source H5 file.
+
+    XML metadata is preferred since it is consistent between products.
+    Neither an XML metadata file nor the H5 source file attributes contain all
+        required information; additional information is extracted from the EOS
+        Metadata Structure in the H5 file whether using an XML or H5 file as the
+        primary metadata source.
+    """
+
+    id: str
+    product: str
+    version: str
+    start_datetime: datetime.datetime
+    end_datetime: datetime.datetime
+    created_datetime: datetime.datetime
+    updated_datetime: Optional[datetime.datetime]
+    horizontal_tile: int
+    vertical_tile: int
+    tile_id: str
+    shape_bounds: Dict[str, Any]
+    densify_factor: Optional[int]
+    xml_href: Optional[str]
+
+    @classmethod
+    def from_xml_and_h5(
+        cls,
+        h5_href: str,
+        read_href_modifier: Optional[ReadHrefModifier] = None,
+        densify_factor: Optional[int] = None,
+    ) -> "Metadata":
+        """Extracts metadata from XML elements and H5 EOS metadata structure"""
+
+        def missing_element(attribute: str) -> Callable[[str], Exception]:
+            def get_exception(xpath: str) -> Exception:
+                return MissingElement(
+                    f"Could not find attribute `{attribute}` at xpath "
+                    f"'{xpath}' at href {xml_href}"
+                )
+
+            return get_exception
```

Comment:
I _may_ be misreading this, but where is `xml_href` defined in this scope?

Reply:
Scope question. Fun! It is available in the enclosing scope. Although it is defined after the nested function, it is defined before the nested function is called, so it is available. I sanity checked this by changing an xpath to be incorrect, and it does throw the error with the correct `xml_href` string content.

I'm open to "better practice" suggestions.

Reply:
Nah this works, just me not scanning Python correctly :-).